src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
+
 public class ArrayList<T> implements List<T> {
 
+    private static final int INITIAL_CAPACITY = 1;
+    private int elementNum = 0;
+    private Object[] list;
+
+    public ArrayList() {
+        list = new Object[INITIAL_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        list = new Object[capacity];
+    }
+
+    private void capacityCheck() {
+        if (elementNum >= list.length) {
+            Object[] enlargedList = new Object[list.length << 1];
+            System.arraycopy(list, 0,enlargedList, 0, elementNum);
+            list = enlargedList;
+        }
+    }
+
+    private void indexCheck(int index) {
+        if (index >= elementNum || index < 0) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        capacityCheck();
+        list[elementNum] = value;
+        elementNum = elementNum + 1;
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexCheck(index);
+        capacityCheck();
+        System.arraycopy(list, index, list, index + 1, elementNum - index);
+        list[index] = value;
+        elementNum = elementNum + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return (T) list[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        if (list[index] == null) {
+            list[index] = value;
+            elementNum = elementNum + 1;
+        } else {
+            list[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Object[] trimmed = new Object[elementNum - 1];
+        T removedElement = (T) list[index];
+        if (index == elementNum - 1) {
```

Comment:
here is my bad, i gave you wrong answer: actual arraylist doesn't trim it's capacity after remove, that's for your information. But you can leave as it is this task will be counted